### PR TITLE
Least authority revisions

### PR DESCRIPTION
**Pull request by mjdatadog**

### CODE REVIEW

**Comment by penandlim** [comment]
Im going to push for having same versions across the release so its not confusing which one to use at deployments.

Also standard math operations should be used (+-/*) in lieu of SafeMath similar to this PR https://github.com/saddle-finance/saddle-contract/pull/393

Version 0.8.17 should be hard enforced in non-interface contracts. For interfaces, we should provide a wide range that is compatible with our compilation versions.

Note that paths are also changed in some of open zeppelin plugins

File: contracts/SwapUtilsV2.sol [hunk 1]
```diff
@@ -0,0 +1,1048 @@
...
+                dP = (dP * (d)) / (xp[j] * (numTokens));
+                // If we were to protect the division loss we would have to keep the denominator separate
+                // and divide at the end. However this leads to overflow with large numTokens or/and D.
+                // dP = dP * D * D * D * ... overflow!
+            }
+            prevD = d;
+            d =
+                (nA * (s)) /
+                (AmplificationUtilsV2.A_PRECISION) +
+                ((dP * (numTokens)) * (d)) /
```
Comment (hannespfeiffer) [inline contracts/SwapUtilsV2.sol:335]: Doesn't that `*(d)` belong one level further out on the paranthesis? Meaning, the whole expression gets multiplied by `d` , not just the `(dP * (numTokens))`

File: contracts/SwapUtilsV2.sol [hunk 2]
```diff
@@ -0,0 +1,1048 @@
...
+                (nA -
+                    ((AmplificationUtilsV2.A_PRECISION) * (d)) /
```
Comment (hannespfeiffer) [inline contracts/SwapUtilsV2.sol:337]: Same here

File: contracts/SwapUtilsV2.sol [hunk 3]
```diff
@@ -0,0 +1,1048 @@
...
+        }
+        c = (c * (d) * (AmplificationUtilsV2.A_PRECISION)) / (nA * (numTokens));
+        uint256 b = s + ((d * (AmplificationUtilsV2.A_PRECISION)) / (nA));
+        uint256 yPrev;
+        uint256 y = d;
+
+        // iterative approximation
+        for (uint256 i = 0; i < MAX_LOOP_LIMIT; i++) {
+            yPrev = y;
+            y = y * (y) + (c) / (y * (2) + (b) - (d));
```
Comment (hannespfeiffer) [inline contracts/SwapUtilsV2.sol:467]: I believe there should be paranthesis around `y * (y) + (c)`

File: contracts/SwapUtilsV2.sol [hunk 4]
```diff
@@ -0,0 +1,1048 @@
...
+                balances[i] = balances[i] + (amounts[i]);
+            } else {
+                balances[i] = balances[i] - amounts[i]; // TODO: remove? "Cannot withdraw more than available";
+            }
+        }
+        uint256 d1 = getD(_xp(balances, multipliers), a);
+        uint256 totalSupply = self.lpToken.totalSupply();
+
+        if (deposit) {
+            return d1 - ((d0) * (totalSupply)) / (d0);
```
Comment (hannespfeiffer) [inline contracts/SwapUtilsV2.sol:614]: Shouldn't this be `((d1 - d0) * (totalSupply)) / (d0)`

File: contracts/SwapUtilsV2.sol [hunk 5]
```diff
@@ -0,0 +1,1048 @@
...
+     * swap fee calculations
+     * @param swapFee swap fee for the tokens
+     * @param numTokens number of tokens pooled
+     */
+    function _feePerToken(uint256 swapFee, uint256 numTokens)
+        internal
+        pure
+        returns (uint256)
+    {
+        return (swapFee * (numTokens)) / (numTokens - (1) * (4));
```
Comment (hannespfeiffer) [inline contracts/SwapUtilsV2.sol:648]: I'd say this should be `(swapFee * (numTokens)) / ((numTokens - (1)) * (4))`

File: contracts/SwapUtilsV2.sol [hunk 6]
```diff
@@ -0,0 +1,1048 @@
...
+        } else {
+            // the initial depositor doesn't pay fees
+            self.balances = newBalances;
+        }
+
+        uint256 toMint;
+        if (v.totalSupply == 0) {
+            toMint = v.d1;
+        } else {
+            toMint = v.d2 - ((v.d0) * (v.totalSupply)) / (v.d0);
```
Comment (hannespfeiffer) [inline contracts/SwapUtilsV2.sol:810]: I think it should be ((v.d2 - v.d0) * (v.totalSupply)) / (v.d0)

**Comment by hannespfeiffer** [comment]
All tests passing

**Review by hannespfeiffer — APPROVED** [review]

**Review by hannespfeiffer — APPROVED** [review]

**Review by penandlim — APPROVED** [review]

**Pull request merged**